rust/src/io/external_manifest/commit.rs

```diff
@@ -0,0 +1,438 @@
+impl CommitHandler for ExternalManifestCommitHandler {
+    /// Get the latest version of a dataset at the path
+    async fn get_latest_version(
+        &self,
+        path: &Path,
+        object_store: &ObjectStore,
+    ) -> std::result::Result<u64, crate::Error> {
+        let version = self
+            .external_manifest_store
+            .get_latest_version(path.as_ref())
+            .await?;
+
+        match version {
+            Some(version) => {
+                let object_store_manifest_path = manifest_path(path, version);
+
+                // external store and object store are out of sync
+                // try to sync them
+                // if sync fails, return error
+                if !object_store.exists(&object_store_manifest_path).await? {
+                    let manifest_location = self
+                        .external_manifest_store
+                        .get(path.as_ref(), version)
+                        .await?;
+
+                    let manifest_path = Path::parse(manifest_location)?;
+                    let staging = make_staging_manifest_path(&manifest_path)?;
+                    object_store.inner.copy(&manifest_path, &staging).await?;
+                    object_store
+                        .inner
+                        .rename(&staging, &object_store_manifest_path)
+                        .await?;
+                    // Also copy for _latest.manifest
+                    object_store.inner.copy(&manifest_path, &staging).await?;
+                    object_store
+                        .inner
+                        .rename(&staging, &latest_manifest_path(path))
+                        .await?;
+                }
+
+                Ok(version)
+            }
+            // Dataset not found in the external store, this could be because the dataset did not
+            // use external store for commit before. In this case, we try to read the latest from
+            // the object store
+            None => read_manifest(object_store, &latest_manifest_path(path))
+                .await
+                .map(|m| m.version),
+        }
+    }
+
+    async fn get_version(
+        &self,
+        path: &Path,
+        version: u64,
+        object_store: &ObjectStore,
+    ) -> Result<String> {
+        let manifest_path = manifest_path(path, version);
+        // found in object store
+        if object_store.exists(&manifest_path).await? {
+            return Ok(manifest_path.to_string());
+        }
+
+        // not found in object store, try to get from external store
+        let physical_path = self
+            .external_manifest_store
+            .get(path.as_ref(), version)
+            .await
+            .map_err(|_| Error::NotFound {
+                uri: manifest_path.to_string(),
+                location: location!(),
+            })?;
+
+        object_store
+            .inner
+            .copy(&Path::parse(physical_path)?, &manifest_path)
+            .await?;
```

Comment:
Is it possible this should have been the latest version and by doing this copy here we prevent a later call to `get_latest_version` from correctly updating `_latest.manifest`?

Reply:
copying to numbered version is always okay as the content is the same.